packages/daimo-api/src/network/viemClient.ts

```diff
@@ -188,18 +190,50 @@ export class ViemClient {
     );
   }
 
-  private async waitForReceipt(hash: Hex) {
+  private async waitForReceipt(hash: Hex, maxWaitMs: number = 10_000) {
+    let receiptStatus = "";
+
     try {
-      const receipt = await this.publicClient.waitForTransactionReceipt({
+      const receiptPromise = this.publicClient.waitForTransactionReceipt({
         hash,
       });
+      const timeoutPromise = new Promise((_, reject) =>
+        setTimeout(
+          () =>
+            reject(
+              new Error(
+                `Timed out after ${maxWaitMs}ms waiting for receipt ${hash}`
+              )
+            ),
+          maxWaitMs
+        )
+      );
+
+      // Wait at most maxWaitMs milliseconds for the receipt
+      const receipt = (await Promise.race([
+        receiptPromise,
+        timeoutPromise,
+      ])) as TransactionReceipt;
       console.log(`[VIEM] waitForReceipt ${hash}: ${JSON.stringify(receipt)}`);
-      if (receipt.status !== "success") {
+
+      receiptStatus = receipt.status;
+      if (receiptStatus !== "success") {
         this.onReceiptError(hash, JSON.stringify(receipt));
       }
     } catch (e) {
       console.error(`[VIEM] waitForReceipt ${hash} error: ${e}`);
       this.onReceiptError(hash, e);
+    } finally {
+      await this.lockNonce.acquireAsync();
+
+      if (receiptStatus === "success") {
+        this.nextNonce += 1;
+        this.pendingTxCount -= 1;
+      } else {
+        this.pendingTxCount -= 1;
+      }
+
+      this.lockNonce.release();
```

Comment:
we might not need lockNonce at all (here or other spot)

Reply:
will leave this in for now and check what `AwaitLock` is actually doing